### DISCUSSION TITLE
15 giugno 2022 - esercizio 1

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/basi-di-dati-1#11] [Esami]
Dato il seguente schema di un DB di una biblioteca:

<div align="center">

LIBRO(<ins>Codice</ins>, Titolo, DataPubblicazione)

SCRITTORE(<ins>Codice</ins>, Nome, Cognome, DataNascita, DataMorte)

AUTORI(<ins>Libro, Autore</ins>)

COPIALIBRO(<ins>CodiceCopia</ins>, CodiceLibro)

</div>

in cui:
- le chiavi di ogni relazione sono sottolineate

- in LIBRO: Codice è il codice assegnato dalla biblioteca ad un libro, Titolo è il titolo del libro, DataPubblicazione è la data in cui è stato pubblicato

- in SCRITTORE: Codice è il codice assegnato dalla biblioteca ad uno scrittore, Nome e Cognome sono nome e cognome dello scrittore, DataNascita è la sua data di nascita, e se è ancora in vita DataMorte è 00/00/0000

- in AUTORI: Libro è il codice assegnato dalla biblioteca ad un libro, Autore è il codice assegnato dalla biblioteca ad uno scrittore, e un libro può avere più autori (più tuple con lo stesso codice libro e diversi codici autore)

- in COPIALIBRO: CodiceCopia è il codice assegnato dalla biblioteca ad una copia, CodiceLibro è il codice assegnato dalla biblioteca ad un libro,

scrivere le seguenti query in algebra relazionale:

1a) Dati dei libri che hanno l'autore o almeno uno degli autori ancora in vita

1b) i dati di tutti i libri pubblicati nel 2021 di cui esiste una sola copia

**Comment by MarcoLinardi** [comment]
![IMG_0286](https://github.com/sapienzastudentsnetwork/bdi2324/assets/123003372/003d0675-f8a1-4aad-99db-9214ed8753ae)

**Reply by MarcoLinardi** [reply]
non ne sono sicuro...

**Reply by IonutCicio** [reply]
Perché la prima l'hai fatta con il complemento? In questo modo hai preso i libri con ***tutti*** gli autori deceduti... bastava solo la prima parte 😉 

Per quanto riguarda la seconda, `CodiceCopia = NULL` non si può ***mai*** verificare, dato che CodiceCopia è la chiave, e la chiave non può essere NULL

> [!NOTE]
> Ho chiamato `COPIASX` e `COPIADX` le tabelle con cui faccio la $\bowtie$ ma mi riferisco sempre a `COPIALIBRO`, è per chiarezza, e all'esame dovrai fare così, altrimenti non puoi distinguere le due tabelle

Quello che andrebbe fatto è 

$\underset{\text{ COPIASX.CodiceCopia } \neq \text{ COPIADX.CodiceCopia }}{\sigma}(\text{COPIASX} \underset{\text{ COPIASX.CodiceLibro } =\text{ COPIADX.CodiceLibro }}{\bowtie} \text{COPIADX})$

In questo modo, con la join metti insieme a coppie tutte le copie dello stesso libro. Con la selezione, rimuovi le righe che hanno lo stesso libro a destra e a sinistra. Le tuple rimanenti avranno lo stesso `CodiceLibro`, ma `CodiceCopia` diverso per destra e sinistra, quindi per forza di cose indicano i libri con più di una copia.

Ora puoi fare il passaggio a complemento e ricavare il risultato finale.

**Reply by MarcoLinardi** [reply]
Scusa nella prima ho fatto la selezione di tutti i libri con autori morti e poi ho li ho tolti da LIBRO, sottraendoli come fa a tornarmi che ho considerato solo autori morti?

**Reply by IonutCicio** [reply]
Perchè nel testo c'è scritto

> se è ancora in **vita**, DataMorte è 00/00/0000

Quindi, quando hai selezionato con DataMorte = 00/00/0000 hai preso quelli in vita, non quelli morti.

**Reply by MarcoLinardi** [reply]
ah so stupido io, grazie

**Comment by atyion** [comment]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/101365532/2bd3e1e1-8481-4528-bb71-6b33f802cd3c)

Spiego il perché della mia risposta a 1b: se facciamo il prodotto cartesiano fra le due tabelle copialibro e selezioniamo i libri che hanno lo stesso codicelibro ma un diverso codicecopia (e facciamo la proiezione su codicelibro) abbiamo il complemento di quello che vogliamo ottenere: la tabella dei libri con più di una copia.

**Reply by IonutCicio** [reply]
LGTM 👍

**Reply by LeoCosta02** [reply]
Chiedeva quelli pubblicati nel 2021 , devi aggiungerlo

**Reply by atyion** [reply]
> Chiedeva quelli pubblicati nel 2021 , devi aggiungerlo

Hai ragione :P

**Comment by IonutCicio** [comment]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/40034429/8d5df7ff-b3df-43dd-bbd2-3671176b78cc)

**Comment by bobbylinux** [comment]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/7393062/24365580-819e-4b00-ae53-0f15defbdf69)

**Comment by sprigatita** [comment]
Non che mi convinca molto...
![photo_5945247654617400992_y](https://github.com/user-attachments/assets/3f99caac-b645-4214-9a1d-0f38379c5f3b)
